docs/source/migration.mdx

```diff
@@ -580,580 +747,456 @@ method.
 
-Apollo Server 3 supports an edge case where the `variables` and
-`extensions` fields inside a `POST` request's JSON-encoded body can be JSON-encoded strings.
+To continue to show a splash page like in Apollo Server 3, write:
 
-Apollo Server 4 requires that within a `POST` request's JSON-encoded body, the `variables` and `extensions` fields must be objects (not doubly-encoded).
+<MultiCodeBlock>
 
-For example, below is a valid query:
+```ts
+import { ApolloServer } from '@apollo/server';
+import { ApolloServerPluginLandingPageLocalDefault, ApolloServerPluginLandingPageProductionDefault }
+  from '@apollo/server/plugin/landingPage/default';
 
-```json
-{
-  “query”: “{ __typename }”, extensions: { “foo”: 1 }
-}
+new ApolloServer({
+  // ...
+  plugins: [
+    process.env.NODE_ENV === 'production'
+      ? ApolloServerPluginLandingPageProductionDefault()
+      : ApolloServerPluginLandingPageLocalDefault({ embed: false })
+  ],
+});
 ```
 
-Whereas this query would be invalid:
-
-```json
-{
-  “query”: “{ __typename }”, extensions: “{ \“foo\”: 1 }”
-}
-```
+</MultiCodeBlock>
 
-### Local landing page defaults to Apollo Sandbox
+## TypeScript-only changes
 
-Apollo Server 4's default development landing page is an embedded Apollo Sandbox, enabling you to run GraphQL operations with no additional CORS setup. If you want to use Apollo Server's default landing page instead, pass `ApolloServerPluginLandingPageLocalDefault({embed: false})` to the plugins of your `ApolloServer` constructor.
+Some changes only affect TypeScript typings, not runtime behavior. For example, we have changed some of the names of TypeScript interfaces to be more clear, and we've changed what packages we use to define other interfaces. (Changes that affect more than just typings (like renaming `GraphQLServiceContext` and `GraphQLServiceConfig`) are described elsewhere.)
 
-### Warning for servers without draining
 
-Versions of Apollo Server 3.2 and above add a "draining" phase to server shutdown, enabling a server to complete in-progress operations before continuing to shut down. Without this draining phase, operations can reach your server even after it has stopped, causing those operations to fail.
+### Improved typing for `context`
 
-In Apollo Server 4, if your server _hasn't_ set up draining and it receives an operation as the server is shutting down, the server logs a warning before failing that operation.
+In Apollo Server 3, you don't tell TypeScript what type your context value is, so there is no compile-time check that the type returned by your `context` function matches the context value type read by your resolvers and plugins. `ApolloServer` does have a generic parameter, but that parameter is the type of the arguments *passed to* your context function (and is specified by your web integration) rather than your app's context value type.
```

Comment:
Tried rewording this to be a little clearer, and I think this is still technically correct, but please correct me if I'm wrong:
```suggestion
In Apollo Server 3, there is no compile-time check that the type your `context` function returns matches the type of your `context`'s value (read by your resolvers and plugins). `ApolloServer` has a generic parameter, but that parameter is the type of the *arguments passed* to your `context` function, _not_  the type of your app's `context` value. 
```

Reply:
It's still accurate. I do think there was some value in saying that you literally never wrote anywhere "the context of my server is MyContext" though my wording may not have been that clear.

Reply:
96401a1d7